SignalMessaging/attachments/SignalAttachment.swift

```diff
@@ -767,10 +781,137 @@ public class SignalAttachment: NSObject {
+            let attachment = SignalAttachment(dataSource : DataSourceValue.emptyDataSource(), dataUTI: dataUTI)
+            attachment.error = .couldNotConvertToMpeg4
+            return (Promise(value: attachment), nil)
+        }
+
+        exportSession.shouldOptimizeForNetworkUse = true
+        exportSession.outputFileType = AVFileTypeMPEG4
+
+        let exportURL = videoTempPath.appendingPathComponent(UUID().uuidString).appendingPathExtension("mp4")
+        exportSession.outputURL = exportURL
+
+        let (promise, fulfill, _) = Promise<SignalAttachment>.pending()
+
+        Logger.debug("\(self.TAG) starting video export")
+        exportSession.exportAsynchronously {
+            Logger.debug("\(self.TAG) Completed video export")
+            let baseFilename = dataSource.sourceFilename
+            let mp4Filename = baseFilename?.filenameWithoutExtension.appendingFileExtension("mp4")
+
+            guard let dataSource = DataSourcePath.dataSource(with: exportURL) else {
+                owsFail("Failed to build data source for exported video URL")
+                let attachment = SignalAttachment(dataSource : DataSourceValue.emptyDataSource(), dataUTI: dataUTI)
+                attachment.error = .couldNotConvertToMpeg4
+                fulfill(attachment)
+                return
+            }
+
+            dataSource.sourceFilename = mp4Filename
+
+            let attachment = SignalAttachment(dataSource: dataSource, dataUTI: kUTTypeMPEG4 as String)
+            fulfill(attachment)
+        }
+
+        return (promise, exportSession)
+    }
+
+    @objc
+    public class VideoCompressionResult: NSObject {
+        @objc
+        public let attachmentPromise: AnyPromise
+
+        @objc
+        public let exportSession: AVAssetExportSession?
+
+        fileprivate init(attachmentPromise: Promise<SignalAttachment>, exportSession: AVAssetExportSession?) {
+            self.attachmentPromise = AnyPromise(attachmentPromise)
+            self.exportSession = exportSession
+            super.init()
+        }
+    }
+
+    @objc
+    public class func compressVideoAsMp4(dataSource: DataSource, dataUTI: String) -> VideoCompressionResult {
+        let (attachmentPromise, exportSession) = compressVideoAsMp4(dataSource: dataSource, dataUTI: dataUTI)
+        return VideoCompressionResult(attachmentPromise: attachmentPromise, exportSession: exportSession)
+    }
+
+    public class func isInvalidVideo(dataSource: DataSource, dataUTI: String) -> Bool {
+        guard videoUTISet.contains(dataUTI) else {
+            // not a video
+            return false
+        }
+
+        guard isInputVideoValidOutputVideo(dataSource: dataSource, dataUTI: dataUTI) else {
+            // found a video which needs to be converted
+            return true
+        }
+
+        // It is a video, but it's not invalid
+        return false
+    }
+
+    private class func isInputVideoValidOutputVideo(dataSource: DataSource?, dataUTI: String) -> Bool {
```

Comment:
Why not just `isValidOutputVideo`?

Reply:
I was matching the convention that we used for image output, but I can rename them both:

`isInputImageValidOutputImage` -> `isValidOutputImage`
`isInputVideoValidOutputVideo ` -> `isValidOutputVideo`